Userland/uniq.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2020, Matthew L. Curry <matthew.curry@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibCore/ArgsParser.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static FILE* get_stream(const char* filepath, const char* perms)
+{
+    FILE* ret;
+
+    if (filepath == nullptr) {
+        if (perms[0] == 'r')
+            return stdin;
+        return stdout;
+    }
+
+    ret = fopen(filepath, perms);
+    if (ret == nullptr) {
+        perror("fopen");
+        exit(1);
+    }
+
+    return ret;
+}
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio rpath wpath cpath", nullptr) > 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    const char* inpath = nullptr;
+    const char* outpath = nullptr;
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(inpath, "Input file", "input", Core::ArgsParser::Required::No);
+    args_parser.add_positional_argument(outpath, "Output file", "output", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    FILE* infile = get_stream(inpath, "r");
+    FILE* outfile = get_stream(outpath, "w");
+
+    char buffers[2][BUFSIZ];
```

Comment:
But what if the line read doesn't fit into `BUFSIZ` bytes?

Reply:
I had used static buffers because sort did, and I figured there was a reason for it (e.g., readline implementation not ready). I've converted to using readline for lines of unlimited length (and opened a pull request on sort to achieve the same -- #3769).

Reply:
It would be nice if we had an API to read a line as `AK::String`. Do we have such a thing somewhere already, perhaps in `Core::IODevice`? If we don't, we probably should have.

Reply:
Hi there,

I wasn't able to find a readline-esque method, unless one's been added in the last few days. I'll be happy to update this utility to use such a method (and the ensuing dependent classes) once such a method appears. I don't think it's correct to add core functionality in this PR for a utility, readline deserves its own discussion.